### ISSUE TITLE
Training results of Ant-v2, Halfcheetah-v2 

### ISSUE THREAD

**Issue by Ending2015a** [apexrl/EBIL-torch#1]
Hello,
Thank you for this amazing project. Could you provide the training results (true rewards) of MuJoCo `Ant-v2` and `Halfcheetah-v2` environments trained with optimal demonstrations? I only see `Hopper-v2` in the original paper.
Besides, the `run_scripts/gen_expert_demos.py` is missing so I can't run your scripts to reproduce the results.

**Comment by Ericonaldo** [comment]
I must admit that my implementation of EBIL can not reach a good performance with the optimal experts on these two envs. 

As you see in the paper, I published the results on HalfCheetah and Ant but they are done with sub-optimal demos (OpenAI baseline version PPO instead of SAC). I think the reason may be the robustness of the training of the energy model with DEEN.

However, a recent work 'Imitation with Neural Density Models' shows that such a density model can reach the optimal performance on these envs. In the essence, they are the same ideas for using a pre-trained density (or energy model) as the reward to train the RL algorithm. The main difference is that they replace the DEEN with a SoTA energy model training method, sliced score matching (SSM). Therefore, I think you can try the new SSM to conduct such experiments instead of DEEN. (Although I don't know how much effort on the finetune since I've tried a bit the method and it does not work that easily, but SSM is definitely a better choice than DEEN)

Best wishes.

**Comment by Ending2015a** [comment]
Ok, thank you for your thorough explanation. So I guess its normal that EBIL cannot solve Ant and HalfCheetah according to your text. I ask this question because I am trying to reproduce the EBIL results on these two envs using my own framework but it doesn't work at all (currently only tested on Hopper and Ant and it worked on Hopper), which may imply using only the energy model as the reward function is not enough. But still it is a great work. thank you.

**Comment by Ericonaldo** [comment]
From my point of view, I can only say that my implementation of EBIL (using DEEN) can not solve Ant :). However as I say, the paper 'Imitation with Neural Density Models' is also a version of EBIL use another training method for the EBM (ssm or [ncfn](https://github.com/ermongroup/ncsn)), and it works as shown in its paper. But he has not published his code until now.

Therefore, I think the essence is not "only the energy model as the reward function is not enough.", but "a good way to train a well EBM is important".

**Comment by Ending2015a** [comment]
I agree that a good way to train the EBM is essential. However, DEEN loss is very similar to the Denoising Score Matching (DSM) loss (only differs by a sigma^2 scale), which Yang Song shows in his paper It can produce competitive results to SSM loss on high-dimensional image generation tasks. So I think there is no much difference between these two losses on training the EBM. 

In addition, I suspect that the improvement of Neural Density Imitation (NDI) does not come from "a well EBM". His ablation study (C.2) shows that there is no significant improvement between training the EBM with 1 expert trajectory and 25 expert trajectories, while I believe training the EBM with only 1 expert trajectory (<1000 expert samples) is definitely not enough. 
Sadly, I can't verify my guess since this paper doesn't provide many implementation details.

**Comment by Ericonaldo** [comment]
"EBM with 1 expert trajectory and 25 expert trajectories", may be explained as that EBM can be trained with 1 expert trajectory (1000 steps) in these envs. But I am also eager to know how he tunes his energy model so well.

**Comment by Ending2015a** [comment]
I mean he also use some additional rewards in his reward function, e.g. log pi(a|s) and f(s, s'), not just the EBM. That's why I said "only EBM is not enough."

**Comment by Ericonaldo** [comment]
See his Table 2, a small lambda can achieve better results. So the additional reward is less useful. Furthermore, log pi is not a reward, but a maximizing entropy term as in SAC.

**Comment by Ending2015a** [comment]
You are right. Then I think I should try SSM first.
